File: spec/lua/test_str_encodings_escaping_enc.lua
local luaunit = require("luaunit")

require("str_encodings_escaping_enc")

TestStrEncodingsEscapingEnc = {}

local function assertUnknownEncoding(expected_encoding, fn)
    luaunit.assertErrorMsgContains("Encoding " .. expected_encoding .. " not supported", fn)
end

function TestStrEncodingsEscapingEnc:test_str_encodings_escaping_enc()
    local r = StrEncodingsEscapingEnc:from_file("src/str_encodings.bin")

    assertUnknownEncoding(
        "ASCII\\\\x",
        function()
            local _ = r.str1.v
        end
    )
    assertUnknownEncoding(
        "UTF-8\\'x",
        function()
            local _ = r.str2.v
        end
    )
    assertUnknownEncoding(
        "SJIS\\\"x",
        function()
            local _ = r.str3.v
        end
    )
    assertUnknownEncoding(
        "IBM437\\nx",
        function()
            local _ = r.str4.v
        end
    )
end

File: spec/javascript/test_str_encodings_escaping_enc.js
var assert = require('assert');
var testHelper = require('testHelper');

// FIXME: fix duplication of `test_str_encodings_escaping_to_s.js` (this is a
// copy of the `validateErr` function from there without comments)
function validateErr(expectedEncoding, err) {
  if (err instanceof RangeError) {
    assert.strictEqual(err.code, 'ERR_ENCODING_NOT_SUPPORTED');
    assert.strictEqual(err.message, 'The "' + expectedEncoding + '" encoding is not supported');
  } else if (Object.getPrototypeOf(err) === Error.prototype) {
    var regex = /^Encoding not recognized: '(.*)' \(searched as: '.*'\)$/;
    var match = err.message.match(regex);
    assert.ok(match !== null, "message [" + err.message + "] does not match regex " + regex);
    assert.strictEqual(match[1], expectedEncoding);
  } else {
    assert.fail("expected " + RangeError.name + " or " + Error.name + ", but got " + err);
  }
  return true;
}

function assertUnknownEncoding(expectedEncoding, fn) {
  assert.throws(fn, validateErr.bind(null, expectedEncoding));
}

testHelper('StrEncodingsEscapingEnc', 'src/str_encodings.bin', function(r, StrEncodingsEscapingEnc_) {
  assertUnknownEncoding(
    "ASCII\\\\x",
    function() {
      r.str1.v;
    }
  );
  assertUnknownEncoding(
    "UTF-8\\'x",
    function() {
      r.str2.v;
    }
  );
  assertUnknownEncoding(
    "SJIS\\\"x",
    function() {
      r.str3.v;
    }
  );
  assertUnknownEncoding(
    "IBM437\\nx",
    function() {
      r.str4.v;
    }
  );
});
